README.md
# Monte

Monte is a simple gem designed to help software engineers and developers answer
the question 'When will it be done?'.

The tool uses the Monte Carlo method to provide a reasonable forecast of the
future based on statistically relevant data from the past.
## Installation

You must have ruby installed to use this tool.

start by installing the gem.

```sh
$ gem install monte
```

## Usage

The following is the simplest version of Monte which should be used when you
have little previous data with which to build your forecast.

```sh
$ monte carlo
 |  \/  |   ___    _ __   | |_    ___
 | |\/| |  / _ \  | '_ \  | __|  / _ \
 | |  | | | (_) | | | | | | |_  |  __/
 |_|  |_|  \___/  |_| |_|  \__|  \___|

Welcome to Monte, a tool to help you answer the question: 'When will the work be done?'


Please answer the following questions
How many tasks/tickets do you have left to complete? 40
Do you have a JIRA csv export to use? no
Enter the smallest number of tasks/tickets you have finished in a week? 2
Enter the largest number of tasks/tickets you have finished in a week? 6
When will you start work (e.g. 28/04/2021) 2020-11-21
How certain are you with regard to the scope of the work? high
How many simulations would you like to run? 10000
┌──────────────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┐
│Forecast Certainty│    5%    │   15%    │   30%    │   50%    │   70%    │   85%    │   95%    │
├──────────────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┤
│  Forecast Date   │2021-02-06│2021-02-06│2021-02-13│2021-02-13│2021-02-20│2021-02-27│2021-03-06│
└──────────────────┴──────────┴──────────┴──────────┴──────────┴──────────┴──────────┴──────────┘
```

Forecasting like this works better when you use real data from the past to model
what the future will look like. To enable this Monte can use a JIRA export of
tasks completed over whatever historic timeframe you would like. Once you have
created your JQL filter ensure that you include the "Resolved" column and then
export as a csv. When running the app you will be given the option to specify
the path of the file and this will then build a distribution of previous
throughputs for use in the simulations. The following is a runthrough using a
test data file.

```sh
  __  __                   _
 |  \/  |   ___    _ __   | |_    ___
 | |\/| |  / _ \  | '_ \  | __|  / _ \
 | |  | | | (_) | | | | | | |_  |  __/
 |_|  |_|  \___/  |_| |_|  \__|  \___|

Welcome to Monte, a tool to help you answer the question: 'When will the work be done?'


Please answer the following questions
How many tasks/tickets do you have left to complete? 40
Do you have a JIRA csv export to use? Yes
what is the absolute file path to the csv file /Users/user/directory/monte/spec/test_data/data.csv
When will you start work (e.g. 28/04/2021) 2020-11-21
How certain are you with regard to the scope of the work? medium
How many simulations would you like to run? 10000
┌──────────────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┐
│Forecast Certainty│    5%    │   15%    │   30%    │   50%    │   70%    │   85%    │   95%    │
├──────────────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┤
│  Forecast Date   │2021-02-06│2021-02-13│2021-02-20│2021-03-06│2021-03-13│2021-03-27│2021-04-10│
└──────────────────┴──────────┴──────────┴──────────┴──────────┴──────────┴──────────┴──────────┘
```


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run
`rake spec` to run the tests. You can also run `bin/console` for an interactive
prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To
release a new version, update the version number in `version.rb`, and then run
`bundle exec rake release`, which will create a git tag for the version, push
git commits and tags, and push the `.gem` file to
[rubygems.org](https://rubygems.org).

## Copyright

Copyright (c) 2020 Andrew Werner. See [MIT License](LICENSE.txt) for further details.


CHANGELOG.md
# Change log

## [v0.2.0] - 2020-11-21

### Added
* Updated the `carlo` command to ask for a JIRA data export that can be used to
  build a distribution of historic throughput based on a JQL filter

## [v0.1.0] - 2020-11-08

### Added
* Added core command: `carlo`


spec/monte_spec.rb
RSpec.describe Monte do
  it "has a version number" do
    expect(Monte::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


spec/spec_helper.rb
require "bundler/setup"
require "monte"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/unit/carlo_spec.rb
# frozen_string_literal: true

require 'monte/commands/carlo'

RSpec.describe Monte::Commands::Carlo do
  it 'executes `carlo` command successfully' do
    title = "\nMonte\n"
    args = {
      backlog: 20,
      split: 1.2,
      runs: 1000,
      start: Date.parse('2020-10-10'),
      throughput: [4, 4, 4]
    }
    result = "#{title}#{Monte::Commands::Carlo::BLURB}#{args[:start] + 6 * 7}\n"
    command = Monte::Commands::Carlo.new
    allow(command).to receive(:ask_questions) { args }
    allow(command).to receive(:create_header) { title }
    allow(command).to receive(:create_table) do |rows|
      rows[0].to_s
    end
    expect do
      command.execute
    end.to output(result).to_stdout
  end
end


lib/monte/cli.rb
# frozen_string_literal: true

require 'thor'

module Monte
  # Handle the application command line parsing
  # and the dispatch to various command objects
  #
  # @api public
  class CLI < Thor
    # Error raised by this runner
    Error = Class.new(StandardError)

    desc 'version', 'monte version'
    def version
      require_relative 'version'
      puts "v#{Monte::VERSION}"
    end
    map %w[--version -v] => :version

    desc 'carlo', 'Runs through a set of questions to generate a forecast for project completion'
    method_option :help, aliases: '-h', type: :boolean,
                         desc: 'Display usage information'
    def carlo
      if options[:help]
        invoke :help, ['carlo']
      else
        require_relative 'commands/carlo'
        Monte::Commands::Carlo.new.execute
      end
    end
  end
end


lib/monte/command.rb
# frozen_string_literal: true

require 'forwardable'

module Monte
  class Command
    extend Forwardable

    def_delegators :command, :run

    # Execute this command
    #
    # @api public
    def execute(*)
      raise(
        NotImplementedError,
        "#{self.class}##{__method__} must be implemented"
      )
    end

    # The external commands runner
    #
    # @see http://www.rubydoc.info/gems/tty-command
    #
    # @api public
    def command(**options)
      require 'tty-command'
      TTY::Command.new(options)
    end

    # The cursor movement
    #
    # @see http://www.rubydoc.info/gems/tty-cursor
    #
    # @api public
    def cursor
      require 'tty-cursor'
      TTY::Cursor
    end

    # Open a file or text in the user's preferred editor
    #
    # @see http://www.rubydoc.info/gems/tty-editor
    #
    # @api public
    def editor
      require 'tty-editor'
      TTY::Editor
    end

    # File manipulation utility methods
    #
    # @see http://www.rubydoc.info/gems/tty-file
    #
    # @api public
    def generator
      require 'tty-file'
      TTY::File
    end

    # Terminal output paging
    #
    # @see http://www.rubydoc.info/gems/tty-pager
    #
    # @api public
    def pager(**options)
      require 'tty-pager'
      TTY::Pager.new(options)
    end

    # Terminal platform and OS properties
    #
    # @see http://www.rubydoc.info/gems/tty-pager
    #
    # @api public
    def platform
      require 'tty-platform'
      TTY::Platform.new
    end

    # The interactive prompt
    #
    # @see http://www.rubydoc.info/gems/tty-prompt
    #
    # @api public
    def prompt(**options)
      require 'tty-prompt'
      TTY::Prompt.new(options)
    end

    # Get terminal screen properties
    #
    # @see http://www.rubydoc.info/gems/tty-screen
    #
    # @api public
    def screen
      require 'tty-screen'
      TTY::Screen
    end

    # The unix which utility
    #
    # @see http://www.rubydoc.info/gems/tty-which
    #
    # @api public
    def which(*args)
      require 'tty-which'
      TTY::Which.which(*args)
    end

    # Check if executable exists
    #
    # @see http://www.rubydoc.info/gems/tty-which
    #
    # @api public
    def exec_exist?(*args)
      require 'tty-which'
      TTY::Which.exist?(*args)
    end

    def table(headers, rows)
      require 'tty-table'
      TTY::Table.new(headers, rows)
    end

    def large_title(title)
      require 'tty-font'
      TTY::Font.new(:standard).write(title)
    end
  end
end


lib/monte/simulation.rb
# frozen_string_literal: true

# The code to run n monte carlo simulations
# args is a hash that must contain the following:
# :backlog
# :split_factor
# :runs
# :low
# :high
module Simulation
  PERCENTILES = [0.05, 0.15, 0.3, 0.5, 0.7, 0.85, 0.95].freeze

  def percentiles(options)
    results = run_simulations(options).sort
    PERCENTILES.map do |percentile|
      index = options[:runs] * (percentile - 1)
      results[index]
    end
  end

  def run_simulations(options)
    estimated_backlog = options[:backlog] * options[:split]
    Array.new(options[:runs]) do |_|
      options[:start] + simulate(
        estimated_backlog,
        options[:throughput]
      ) * 7
    end
  end

  def simulate(backlog, throughput, result = 0)
    return result if backlog <= 0

    simulate(
      backlog - throughput.sample,
      throughput,
      result + 1
    )
  end
end


lib/monte/commands/carlo.rb
# frozen_string_literal: true

require 'date'
require_relative '../command'
require_relative '../simulation'
require_relative '../csv_data'

module Monte
  module Commands
    # Runs Monte Carlo Simulation to estimate how long a piece of work will take
    class Carlo < Monte::Command
      include Simulation
      include CSVData
      BLURB = %(Welcome to Monte, a tool to help you answer the question: 'When will the work be done?'\n\n
Please answer the following questions\n)
      CERTAINTY = { 'certain' => 1.0, 'high' => 1.2, 'medium' => 1.5, 'low' => 1.8 }.freeze
      RUNS = { '10000' => 10_000, '1000' => 1000, '500' => 500 }.freeze
      HEADERS = ['Forecast Certainty', '5%', '15%', '30%', '50%', '70%', '85%', '95%'].freeze
      PERCENTILES = [0.05, 0.15, 0.3, 0.5, 0.7, 0.85, 0.95].freeze

      def execute
        puts(create_header, BLURB)
          .then { |_| ask_questions({}) }
          .then { |answers| percentiles(answers) }
          .then { |pcentile| create_table(pcentile) }
          .then { |table| puts(table) }
      end

      def create_table(rows)
        table(HEADERS, [rows.prepend('Forecast Date')])
          .render(:unicode, alignment: [:center])
      end

      def create_header
        large_title('Monte')
      end

      def ask_questions(options)
        ask_backlog(options)
          .then { |opts| ask_start(opts) }
          .then { |opts| ask_throughput(opts) }
          .then { |opts| ask_split(opts) }
          .then { |opts| ask_runs(opts) }
      end

      def ask_backlog(options)
        options.merge(backlog: prompt.ask(
          'How many tasks/tickets do you have left to complete?',
          required: true,
          convert: :int
        ))
      end

      def ask_split(options)
        options.merge(split: prompt.select(
          'How certain are you with regard to the scope of the work?',
          CERTAINTY
        ))
      end

      def ask_start(options)
        options.merge(start: prompt.ask('When will you start work (e.g. 28/04/2021)') do |q|
                               q.required true
                               q.default Date.today
                               q.convert ->(input) { Date.parse(input.to_s) }
                             end)
      end

      def ask_throughput(options)
        data_exists = prompt.yes?('Do you have a JIRA csv export to use?', required: true)
        if data_exists
          path = prompt.ask('what is the absolute file path to the csv file', required: true)
          throughput = historic_throughput(path)
        else
          low = prompt.ask('Enter the smallest number of tasks/tickets you have finished in a week?', convert: :int)
          high = prompt.ask('Enter the largest number of tasks/tickets you have finished in a week?', convert: :int)
          throughput = (low..high).to_a
        end
        options.merge(throughput: throughput)
      end

      def ask_runs(options)
        options.merge(
          runs: prompt.select('How many simulations would you like to run?', RUNS)
        )
      end
    end
  end
end
